src/lib.rs

```diff
@@ -1384,6 +1369,165 @@ fn digraph_all_simple_paths(
     Ok(result)
 }
 
+fn weight_callable(
+    py: Python,
+    weight_fn: &Option<PyObject>,
+    weight: &PyObject,
+    default: f64,
+) -> PyResult<f64> {
+    match weight_fn {
+        Some(weight_fn) => {
+            let res = weight_fn.call1(py, (weight,))?;
+            res.extract(py)
+        }
+        None => Ok(default),
+    }
+}
+
+/// Find the shortest path from a node
+///
+/// This function will generate the shortest path from a source node using
+/// Dijkstra's algorithm.
+///
+/// :param PyGraph graph:
+/// :param int source: The node index to find paths from
+/// :param int target: An optional target path to find the path
+/// :param weight_fn: An optional weight function for an edge. It will accept
+///     a single argument, the edge's weight object and will return a float which
+///     will be used to represent the weight/cost of the edge
+/// :param float default_weight: If ``weight_fn`` isn't specified this optional
+///     float value will be used for the weight/cost of each edge.
+/// :param bool as_undirected: If set to true the graph will be treated as
+///     undirected for finding the shortest path.
+///
+/// :return: Dictionary of paths. The keys are destination node indices and
+///     the dict values are lists of node indices making the path.
+/// :rtype: dict
+#[pyfunction(default_weight = "1.0", as_undirected = "false")]
+#[text_signature = "(graph, source, /, target=None weight_fn=None, default_weight=1.0)"]
+pub fn graph_dijkstra_shortest_paths(
+    py: Python,
+    graph: &graph::PyGraph,
+    source: usize,
+    target: Option<usize>,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+) -> PyResult<PyObject> {
+    let start = NodeIndex::new(source);
+    let goal_index: Option<NodeIndex> = match target {
+        Some(node) => Some(NodeIndex::new(node)),
+        None => None,
+    };
+    let mut paths: HashMap<NodeIndex, Vec<NodeIndex>> = HashMap::new();
+    paths.insert(start, vec![start]);
```

Comment:
I don't quite follow the format of the `paths` mapping. Is it a dict of "targets" and their paths to that node? So for graph = `0 -> 1 -> 2` and `start = 0`, would it end up: `{0: [0], 1: [0, 1], 2: [0, 1, 2]`?

Reply:
Yeah, that's it. But we remove the 0 field being added here in the output dict below at L1434. This needs to be in the dict we pass to `dijkstra::dijkstra` because the dijkstra function uses the values in the paths dict on each iteration so without it there it would fail. 

Basically this line is just a starter for building the path dict for the children of the source node so that in: https://github.com/Qiskit/retworkx/pull/162/files/42eb1656062122528ab2051f6cfc23126d630aa5#diff-35666815bfec005ea9e5e2774a0cee5e97ea7af1ee9db63f254f8bfe7d59256aR138-R147 the paths include 0.

Reply:
Actually writing this and adding docstrings for this function. I'm thinking I should just move this line into the `dijkstra::dijkstra()` function. So instead of expecting that users add an entry to the HashMap before passing it to the function we add it before starting the traversal. What do you think?

Reply:
Yes that makes sense to me. It seems that this piece is part of the dijsktra algorithm itself, rather than a required input.

Reply:
Ok, done in: https://github.com/Qiskit/retworkx/pull/162/commits/7eab33db66f81daad36223ee2730265660c8c7e7